### ISSUE TITLE
RedisValue: type mismatch

### ISSUE THREAD

**Issue by naighes** [StackExchange/StackExchange.Redis#2418]
Hello!
I'm experiencing an issue with `RedisValue` underlying type.
So, I'm flushing some data like the following:

```
RedisValue someInt = 12;
db.HashSetAsync(key, new[] {
    new HashEntry("some_int", someInt)
    // ...
});
```

The property `IsInteger()` prints `true` for `someInt`.
The problem occurs when attempting to read back the set:

```
var entry = await db.HashGetAllAsync(key);
var someInt = entry.FirstOrDefault(x => x.Name == "some_int").Value;
```

Here `IsInteger()` returns `false`, and I cannot sort out why.
Furthermore any cast to `Int32` just fails (e.g. `var x = (Int32)someInt;`).
Any hint?

**Comment by naighes** [comment]
> What does `someInt.ToString()` show? I'm interested in whether the value is simply now what we expect
It prints "12", which is correct.
But look at this:

```
foreach (var e in entry) {
    System.Diagnostics.Debug.WriteLine($"{e.Name} : {e.Value.ToString()}");
}
```
It prints:
```
some_int : 12
```

Instead, this:
```
var v = entry.FirstOrDefault(x => x.Name == "some_int").Value.ToString();
```
...well, it prints an empty string for `v`.

**Comment by mgravell** [comment]
That suggests the match wasn't found by name. I'll see what I can find.

**Comment by mgravell** [comment]
I have a repro, but I don't get the empty value - I get a 12, but where IsInteger reports false; I think this is a documentation error, as this isn't intended to mean "can be interpreted as", but instead means "the underlying value being used here is..."

It should work correctly via `someInt.TryParse(out int)`, remembering to check the `bool` return value.

**Comment by mgravell** [comment]
See also: ^^^

**Comment by mgravell** [comment]
(note: if this fixes things, please don't close this - I will add a documentation fix in that branch and add a PR)

**Comment by mgravell** [comment]
Also note for me: check that we have `==` operators (left and right) for: `string` and `int` - we can avoid some things here.

**Comment by NickCraver** [comment]
Fixed in #2420 - closing out to cleanup :)
